lib/utils.js

```diff
@@ -51,9 +72,9 @@ async function bundleFeeds(feeds, type) {
     }
 }
 
-async function upload(sink, fileName, content) {
+async function upload(sink, fileName, content, retries = 0) {
     try {
-        await sink.set(fileName, content);
+        await pRetry(() => sink.set(fileName, content), { retries });
```

Comment:
I understand we are trying to catch errors and fix a bug which happens on writing/reading files to GCS, but the amount of try/catch and error throwing/wrapping worry me a bit.

What will actually happen here if we look at what the underlaying code can reject/throw?

I might be wrong but I can see one issue here: 
* The GCS sink has a [ready check which checks if a bucket exist](https://github.com/asset-pipe/asset-pipe-sink-gcs/blob/master/lib/sink.js#L149). This is a promise which resolves or rejects. 
* The ready check is then [checked when one try to set a file](https://github.com/asset-pipe/asset-pipe-sink-gcs/blob/master/lib/sink.js#L205). If the bucket does not exist one will get an rejection. 
* If a rejection happens the above code will retry the set operation and it does so with default settings of retry which to me looks like will result in a retries for 17 minutes and 3 seconds before a rejection is reached. 
* First then will we get an rejection which we will catch and use to send a http error status code. Default node.js response timeout is 2 minutes so the client requesting this will have timed out long before we are done retrying...

If a bucket does not exist in GCS, is this the behaviour we want when starting the server / calling the http enpoint triggering this?

Reply:
That's a good point.

We should probably make sure to reject with an `AbortError` somehow if the sink is not `ready`, as we really don't want to try any calls at all if that's somehow not in a good state.

We might want to consider putting an `await` for `ready` in the asset-server so that it can handle that separately (probably go down in our case)

Reply:
I think the following:

- The sink should check if bucket exist on construction (when the app starts). One can then have two approaces if a bucket does not exist; create it, flag the app not ready to receive traffic while creating it and when done creating the bucket flag the app as ready and go on. Or throw an error so the app wont stat or its flagged so the app can run but not get into a ready stage to receive traffic. Then it will not be any need to check for readyness in each method of the sink.
- The method in the sink does also throw if arguments to the methods is missing and the same retry logic will happen. Move the retry logic to the sink encapsulating only the calls to the GCS. The sinks is an abstraction to create a common API for different backends and also account for weirdness in them. By having the retry logic here, it will also apply to if one use the mem sink. Which does not make sence. If done so, we will not retry if an argument on the method is missing. If a retry rails, it can throw a boom error.
- The values used for the arguments comes from the http request. These should be validated on request level and if found to be missing, the request should be terminated before we reach this fare down in the application. In theory there should never come to a situation where the method in the sink errors due to missing arguments. Though, its nice that they throw, but it will be rare that they do. If such an rare error happen, it can surface all the way up and be wrapped as a generic error in the Express error handler which should boom wrap all non boom errors. Iow; no need to boom wrap errors here.

By doing so I think we can remove some error wrapping here, we don't have to introduce a consept of custom error objects for each fail scenario and we do retries only on what we want to retry on. And we will know if GCS are in a state to be written to upon start of the app and not on request.

Reply:
Yup, I'm all about this. Nice approach. Much better to encapsulate the retries in the GCS sink etc. Will hop to it. 

From point 1. flag and create bucket vs. throw if not exist, which do we prefer? I'm leaning toward flag and create.

Reply:
>The sink should check if bucket exist on construction (when the app starts).

An issue with this is that constructors cannot be asynchronous, we need to be able to wait for the network call.

We can work around it in our app by awaiting `ready` if it's exposed, but should the different methods on the sink not care if it's ready or not?

>Move the retry logic to the sink encapsulating only the calls to the GCS.

+100

>In theory there should never come to a situation where the method in the sink errors due to missing arguments. Though, its nice that they throw, but it will be rare that they do. If such an rare error happen, it can surface all the way up and be wrapped as a generic error

While I agree in theory, I would much rather have very descriptive and detailed error messages now, as we have issues in production without any error messages at all. I'm all for maybe removing them in the future once we've squashed the bugs we have, but I'd rather have too descriptive errors than generic ones up until that point

Reply:
yeah, also a good point. `.ready` or remove the use of classes ;)

Reply:
> An issue with this is that constructors cannot be asynchronous, we need to be able to wait for the network call.

There are several patterns to deal with this;
* The constructor can be an eventemitter and emit an event when ready.
* It can take a callback which if fired on ready or rejection.
* It can be a method on the construcor which one need to pull.

That we have async await does not eliminate usage of other tools and patterns in our toolbox.

Reply:
Yes, but we should avoid using too many patterns at once.

That said, I'm all for emitting a `ready` event in this case - seems like nice, idiomatic node code

Reply:
And maybe `error` if it fails/times out?

Reply:
>  but should the different methods on the sink not care if it's ready or not?

What is ready? 

For me ready in the sink is that one have managed to authenticate to GCS and checked that one have the buckets one need to write stuff later on. Preferly maybe checked that each bucket have correct read/write settings. Until that is checked, the app are in my opinion, not ready. If we can not log on or read write to the buckets we need this app are in such a state that it can not do much.

I don't think each methods need to have a check for such readyness.

Then, in flight a lot of things can go bonkers. Everything from GCS being down to broken network somewhere on the way or even someone deleted the buckets we write to through a second interface. These things are other types of errors which each method should account for handling, but they are not part of an ready check in my opinion.

Reply:
> From point 1. flag and create bucket vs. throw if not exist, which do we prefer? I'm leaning toward flag and create.

I do also prefer flag and create here. Saves us from adding manual steps for creating buckets if one spinn up a new installation.

Reply:
I disagree. I think the access token used should have the least amount of privileges possible, and creating a bucket is not a permission it needs.

The token we have in production at FINN does not have permission to create a bucket, so it will just throw. I think this is a good practice - read/write access to a specific bucket is enough IMO.

I might agree if we add an option the the sink (`createMissingBucket`, or something), but I don't think we should encourage over-privileged access tokens by default